### PR TITLE
gh-121141: add support for `copy.replace` to AST nodes

### PR DESCRIPTION
**Pull request by picnixz**
cc @JelleZijlstra @serhiy-storchaka 

I eventually had enough time for this one, but I had a question on the implementation and efficiency. Assume that I have the following:

```python
node = ast.Name(id='a')
node.foo = 1
node.bar = 2
```

I cannot do `ast.Name(id='a', foo=...)` in 3.15 (it's deprecated) and this forbids me to call the constructor with the keyword arguments passed to `copy.replace`. So, `copy.replace(node, id='b', foo=0)` with `kwargs = {'id': '5', 'foo': 0}` is implemented as follows:

1. Create `node2 = ast.Name(id='b', ctx=node.ctx)` (this pops 'id' from `kwargs`, or more generally, all fields known to the AST class, but not those extra fields).
2. Do `node2.__dict__ |= kwargs` (i.e., you make the changes on the extra fields).
3. Now `node2.id == 'b'` and `node2.foo = 0` but `node2` has **no** attribute ``bar`` yet because the constructor did not carry this information.
4. So, I need again to do something like `PyDict_Merge(node2.__dict__, node.__dict__, 0)` so that all remaining values from `node.__dict__` that were neither obtained in 1 or 2 are not set.

Ideally, I'd like to create a function that can add me whatever attributes I want without warning me but I'm not sure of the refactoring. So for now, I left that logic separate. There is a separate routine that checks whether you can replace a field or not (e.g., `copy.replace(node, unknown=1)` would fail because the node has no attributes named *unknown* for now).

----

I also fixed `ast.compare` because I may *lack* attributes or fields, and comparing two objects that lack attributes or fields would still be correct IMO. If you want me to make another PR for that, I can cherry pick the commit (it's the first one).

---

EDIT: simple wording question but should it be "add support for [...] to AST nodes" or "for AST nodes"? (or whatever else?)

* Issue: gh-121141

----
📚 Documentation preview 📚: https://cpython-previews--121162.org.readthedocs.build/

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
I think that it should only accept fields known to the AST class.

**Comment by picnixz** [comment]
> I think that it should only accept fields known to the AST class.

I can live with that assumption but I need to update the docs then. However, what about constructions that add parent's back references. I would expect such nodes to be able to copy themselves but only change one single attribute (like a sibling) so that's why I only allowed the fields in `__dict__` to be mutated.

By the way, is there something I should have taken care of because of singletons such as `ast.Load`? Now that I see my tests, I'm not sure if a special treatement is needed.

**Comment by serhiy-storchaka** [comment]
AFAIK, all other `__replace__` implementations only allow to change the fixed set of fields, not arbitrary instance attributes.

**Comment by picnixz** [comment]
Actually, I assumed that AST nodes were somewhat similar to simple namespaces, which allows for arbitrary addition I think:

https://github.com/python/cpython/blob/6b280a84988ca221b5bdc1077a914e873790cce5/Objects/namespaceobject.c#L227-L232

But if you think we should restrict to fields known to the class, I can easily remove the code.

**Comment by JelleZijlstra** [comment]
I feel AST nodes are more like dataclasses. For dataclasses, unrecognized arguments are rejected by `copy.replace` and non-field attributes on the original object are lost.

```
>>> from dataclasses import dataclass
>>> 
>>> @dataclass
... class A:
...     a: int
...     
>>> from copy import replace
>>> a = A(1)
>>> a.b = 42
>>> a2 = replace(a, a=2)
>>> a2
A(a=2)
>>> a2.b
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    a2.b
AttributeError: 'A' object has no attribute 'b'
>>> a.b
42
>>> replace(a, b=5)
Traceback (most recent call last):
  File "<python-input-10>", line 1, in <module>
    replace(a, b=5)
    ~~~~~~~^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/copy.py", line 294, in replace
    return func(obj, **changes)
  File "/Users/jelle/py/cpython/Lib/dataclasses.py", line 1630, in _replace
    return self.__class__(**changes)
           ~~~~~~~~~~~~~~^^^^^^^^^^^
TypeError: A.__init__() got an unexpected keyword argument 'b'
```

**Comment by picnixz** [comment]
Just after I sent my comment, I actually thought that maybe I should reject them since we anyway reject arbitrary keyword arguments in the constructor, maybe it's indeed closer to dataclasses.

I convinced myself as well and since we reached a consensus, I'll remove this. I'll extend the `validate` function (good thing I did this).

**Comment by serhiy-storchaka** [comment]
`SimpleNamespace` constructor accepts arbitrary keyword arguments, so `__replace__()` does the same. All names are valid for `SimpleNamespace`, but other objects are more picky.

**Review by serhiy-storchaka — COMMENTED** [review]
I think that the implementation can be simpler.

* Get fields (`PyObject_GetOptionalAttr((PyObject*)Py_TYPE(self), state->_fields, &fields)`).
* Iterate them, and get corresponding values (`PyDict_GetItemRef(dict, name, &value)`).
* Add corresponding name-value pair to the new dictionary.
* Update that dictionary with the keyword arguments dictionary.
* Create the object of the same type by passing a new dictionary as keyword arguments.

Validation is performed in the constructor.

Equivalent Python code:
```py
def __replace__(self, /, **kwargs):
    newkwargs = {}
    for name in self._fields:
        value = getattr(self, name)
        newkwargs[name] = value
    newkwargs.update(kwargs)
    return type(self)(**kwargs)
```

**Comment by picnixz** [comment]
> I think that the implementation can be simpler.

Yes, that's what I decided after our discussion from yesterday. Currently the construction only warns about extra keyword arguments. For now, I'll leave that warning but this also means that the arguments *are* replaced, even though it is not wanted (for now). For 3.15 and later, the tests should be changed, so I added some comments but is there a more rigourous way?

File: Parser/asdl_c.py [hunk 1]
```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
...
+    struct ast_state *state = get_ast_state();
+    if (state == NULL) {
+        return NULL;
+    }
+
+    PyObject *result = NULL;
+    // known AST class fields
+    PyObject *fields = NULL;
+    // current instance dictionary
+    PyObject *__dict__ = NULL;
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: It is better to not use names with double underscores in C. They are reserved for compiler. Use just "dict".

File: Parser/asdl_c.py [hunk 2]
```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
...
+        for (Py_ssize_t i = 0; i < numfields; i++) {
+            if (!(key = PySequence_GetItem(fields, i))) {
+                goto cleanup;
+            }
+            if (PyDict_GetItemRef(__dict__, key, &value) < 0) {
+                goto cleanup;
+            }
+            if (!value) {
+                break;
+            }
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: This may be surprising. If any field is absent in `__dict__`, all following names will be silently skipped.

File: Parser/asdl_c.py [hunk 3]
```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
...
+    if (!(payload = PyDict_New())) {
+        goto cleanup;
+    }
+    if (fields) {
+        Py_ssize_t numfields = PySequence_Size(fields);
+        if (numfields == -1) {
+            goto cleanup;
+        }
+        for (Py_ssize_t i = 0; i < numfields; i++) {
+            if (!(key = PySequence_GetItem(fields, i))) {
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: Do not use assignment in expression without need.
```suggestion
            key = PySequence_GetItem(fields, i);
            if (key == NULL) {
```
This is only justified in the `while` or `for` condition, in the `else if` condition, and maybe in few other cases.

File: Parser/asdl_c.py [hunk 4]
```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
...
+        goto cleanup;
+    }
+    if (!(result = type->tp_new(type, empty_tuple, payload))) {
+        Py_CLEAR(result);
+        goto cleanup;
+    }
+    if (ast_type_init(result, empty_tuple, payload) < 0) {
+        Py_CLEAR(result);
+        goto cleanup;
+    }
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: The same suggestion about assignment in expression.

And why not simply use `PyObject_Call((PyObject *)type, empty_tuple, payload)`?

File: Parser/asdl_c.py (hunk 3 above)
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: The old value of `key` is leaked.

File: Parser/asdl_c.py [hunk 5]
```diff
@@ -1132,13 +1132,93 @@ def visitModule(self, mod):
...
+    if (fields) {
+        Py_ssize_t numfields = PySequence_Size(fields);
+        if (numfields == -1) {
+            goto cleanup;
+        }
+        for (Py_ssize_t i = 0; i < numfields; i++) {
+            if (!(key = PySequence_GetItem(fields, i))) {
+                goto cleanup;
+            }
+            if (PyDict_GetItemRef(__dict__, key, &value) < 0) {
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: The old value of `value` is leaked.

File: Lib/ast.py [hunk 6]
```diff
@@ -449,8 +451,14 @@ def _compare_fields(a, b):
...
-            a_field = getattr(a, field)
-            b_field = getattr(b, field)
+            a_field = getattr(a, field, sentinel)
+            b_field = getattr(b, field, sentinel)
+            if a_field is sentinel and b_field is sentinel:
+                # both nodes are missing a field at runtime
+                return True
+            if a_field is sentinel or b_field is sentinel:
+                # one of the node is missing a field
+                return False
```
Comment (serhiy-storchaka) [inline Lib/ast.py]: Why is this needed? Is it related to `__replace__`?

File: Parser/asdl_c.py (hunk 5 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Right!  moved it to the cleanup part but forgot to keep the non-exceptional cases (had it in the first version actually...)

File: Parser/asdl_c.py (hunk 4 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Err... I wasn't sure which one is the most efficient way. By calling them separately I thought it would reduce the number of checks but if that's the standard way to do I'll do it (thanks for the advice!)

File: Parser/asdl_c.py (hunk 2 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Oh right, I forgot to remove that one. It was used before because of positional arguments but now it's no more needed. Thanks!

File: Parser/asdl_c.py (hunk 3 above)
Comment (picnixz) [reply Parser/asdl_c.py]: I assume that 
```C
  if (!(empty_tuple = PyTuple_New(0))) {
      goto cleanup;
  }
```

should also be rewritten as 

```C
empty_tuple = PyTuple_New(0);
if (empty_tuple == NULL) {
    goto cleanup;
}
```

File: Parser/asdl_c.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Parser/asdl_c.py]: This code does not use `__new__` and `__init__` if they are overridden in a subclass. Unless you have a reason for such behavior, it is better to just call `type`.

File: Lib/ast.py (hunk 6 above)
Comment (picnixz) [reply Lib/ast.py]: The part on the 'attributes' is indeed needed but actually, it's not a `return True` that I need but a `continue` here.

The reason why it's needed is because if I write something like:
```
import ast
ast.compare(ast.Tuple(), ast.Tuple(), compare_attributes=1)
```

I get

```sh
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    ast.compare(ast.Tuple(), ast.Tuple(), compare_attributes=1)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python/cpython/Lib/ast.py", line 486, in compare
    if compare_attributes and not _compare_attributes(a, b):
                                  ~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/lib/python/cpython/Lib/ast.py", line 472, in _compare_attributes
    a_attr = getattr(a, attr)
AttributeError: 'Tuple' object has no attribute 'lineno'
``` 

And similarly, the same error would be there if I remove a field at runtime. I can cherry pick the fix (and the fix of the fix) and make a separate PR if you prefer though.

File: Parser/asdl_c.py [hunk 7]
```diff
@@ -1148,62 +1148,70 @@ def visitModule(self, mod):
...
+            rc = PyDict_GetItemRef(dict, key, &value);
+            if (rc < 0) {
+                Py_DECREF(key);
                 goto cleanup;
             }
-            if (!value) {
-                break;
+            if (rc == 0) {
+                Py_DECREF(key);
+                Py_DECREF(value);
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: `value` is always NULL here.

Actually, you can even use `value == NULL` instead of `rc == 0` if you prefer.

File: Parser/asdl_c.py [hunk 8]
```diff
@@ -1148,62 +1148,70 @@ def visitModule(self, mod):
...
         goto cleanup;
     }
     if (fields) {
         Py_ssize_t numfields = PySequence_Size(fields);
         if (numfields == -1) {
             goto cleanup;
         }
+        int rc;
+        // for iterating over 'fields' and 'kwargs'
+        PyObject *key = NULL, *value = NULL;
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: Just move this in the loop block.

File: Parser/asdl_c.py (hunk 7 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Wait, I thought I removed that one.

Comment (picnixz) [reply Parser/asdl_c.py]: Ah! I directly edited the C-file... which got overwritten by the generator...

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

You even made the changes I thought about but didn't write so as not to be too picky. We think alike.

File: Lib/ast.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/ast.py]: Is this a bug? Should this change be backported? If yes, then it needs a separate issue and PR.

Comment (picnixz) [reply Lib/ast.py]: I think so IMO. I'll make a separate issue for that then! (I don't think it should be backported since `ast.compare` was introduced in 3.14 (I think?). If the feature made it to 3.13, then it should be backported)

**Comment by picnixz** [comment]
> You even made the changes I thought about but didn't write so as not to be too picky. We think alike. 

Don't worry about pickiness since I'm also quite picky. I always welcome feedbacks, especially when it comes to code that could be used by other people than me.

File: Lib/ast.py (hunk 6 above)
Comment (picnixz) [reply Lib/ast.py]: See https://github.com/python/cpython/issues/121210 (PR: https://github.com/python/cpython/pull/121211).

File: Lib/test/test_ast.py [hunk 9]
```diff
@@ -1199,6 +1216,247 @@ def test_copy_with_parents(self):
...
+    def test_replace_native(self):
+        for klass in self.iter_ast_classes():
+            fields = set(klass._fields)
+            with self.subTest(klass=klass, fields=fields):
+                # use of object() to ensure that '==' and 'is'
+                # behave similarly in ast.compare(node, repl)
+                old_value, new_value = object(), object()
+
+                # check shallow copy
+                node = klass(**dict.fromkeys(fields, old_value))
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py]: Shouldn't this use a different `old_value` for each field?

File: Lib/test/test_ast.py [hunk 10]
```diff
@@ -1199,6 +1216,247 @@ def test_copy_with_parents(self):
...
+            fields = set(klass._fields)
+            with self.subTest(klass=klass, fields=fields):
+                node = klass(**dict.fromkeys(fields))
+                # forbid positional arguments in replace()
+                self.assertRaises(TypeError, copy.replace, node, 1)
+                self.assertRaises(TypeError, node.__replace__, 1)
+
+    def test_replace_native(self):
+        for klass in self.iter_ast_classes():
+            fields = set(klass._fields)
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py:1254]: Also check the attributes

File: Lib/test/test_ast.py [hunk 11]
```diff
@@ -1199,6 +1216,247 @@ def test_copy_with_parents(self):
...
+            x = 0
+            y = 1
+            _attributes = ('x', 'y',)
+
+        node = MyNode()
+        self.assertEqual(node.x, 0)
+        self.assertEqual(node.y, 1)
+
+        y = object()
+        # Currently, we emit a warning (and thus the changes are committed)
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py]: Should we raise an error here immediately instead? It doesn't make sense to add new functionality and immediately deprecate part of it.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks! My main feedback is that I think `copy.replace` should start throwing an error now rather than in 3.15 for missing required fields and for unrecognized fields.

File: Lib/test/test_ast.py (hunk 11 above)
Comment (picnixz) [reply Lib/test/test_ast.py]: Well.. it's mainly because of the warning saying that it will become an error in 3.15 (but not before). I didn't want to change the constructor =/ and Serhiy suggested to leave the verification to the constructor instead of checking it inside `copy.replace`.

File: Lib/test/test_ast.py (hunk 9 above)
Comment (picnixz) [reply Lib/test/test_ast.py]: Mmh, technically we don't really care here. It's just to avoid the warning that I did that (mainly to fill up the values that the constructor expects). here, it's mainly to check that `copy.replace(repl)` (when you do nothing) does nothing else. But I'll make them all independent just in case.

File: Lib/test/test_ast.py (hunk 10 above)
Comment (picnixz) [reply Lib/test/test_ast.py:1254]: Ah yes, good catch. The attributes are not copied actually, so I'll do it. Thanks!

File: Lib/test/test_ast.py (hunk 11 above)
Comment (picnixz) [reply Lib/test/test_ast.py]: Are you allowed to change the deprecation warning that was introduced in 3.13 (https://github.com/python/cpython/issues/105858) or is it not possible anymore?

**Comment by picnixz** [comment]
Apart from the question on whether to raise an exception or not now, I have fixed the case where attributes are not copied actually. But I would like to know whether you prefer that I *only* use positional arguments for fields and keyword arguments for attributes or if you are happy with passing everything as keyword arguments.

File: Lib/test/test_ast.py (hunk 11 above)
Comment (JelleZijlstra) [reply Lib/test/test_ast.py]: I'm not sure why we'd need to change the deprecation warning. The behavior we deprecated was existing behavior. Here we're adding a new feature, so it doesn't have anything to do with the deprecation.

Comment (picnixz) [reply Lib/test/test_ast.py]: Ah so you suggest raising the error inside `__replace__` and not inside the constructor call. I can (re-)add that check!

**Comment by picnixz** [comment]
@JelleZijlstra I changed the implementation but I'm still leaving one place where I'm not sure whether you want me to check or not the keyword arguments, namely, when a field is missing from and I do `copy.replace(node)`, currently it only emits a warning. Should I leave it like that or should I also check whether the instance being currently replaced is lacking fields or attributes? (IMO, it would defeat the purpose of replacing attributes/fields on partial nodes, and I  would rather not have a warning *at all* for such cases, but I don't know if you want `__replace__` to only be allowed on complete AST nodes).

**Comment by picnixz** [comment]
I added two commits in which the optional attributes are checked as follows:

I have other questions as well and I'm not sure if we should first patch them before continuing this.

- The allowed attributes are hard-coded and thus even if I specify them in sub-classes, I get a warning. Is it fine?

    ```python
	>>> import ast
	>>> class A(ast.AST):
	...     _attributes = ('a',)
	...     a = None
	...
	>>> A(a=1)
	<python-input-6>:1: DeprecationWarning: A.__init__ got an unexpected keyword argument 'a'. Support for arbitrary keyword arguments is deprecated and will be removed in Python 3.15.
	  A(a=1)
	<__main__.A object at 0x7f021ddb08d0>
    ``` 

- Because of this behaviour, I still have a warning being emitted. However, `copy.replace` correctly raises exceptions whenever a field that is not declared explicitly on the class is missing or not recognized. 

- Because attributes are sometimes optional and sometimes not, I do not have a way to check whether they should be supplied or not easily. I have a little trick where I try to find the expected attributes on the current instance, and if they exist, I assume that they do not need to be supplied again. In addition, it appears that "required attributes" cannot be deleted:

    ```python
	>>> ast.Expr(1).end_lineno is None
	True
	>>> del ast.Expr(1).end_lineno
	Traceback (most recent call last):
	File "<python-input-11>", line 1, in <module>
	  del ast.Expr(1).end_lineno
	      ^^^^^^^^^^^^^^^^^^^^^^
	AttributeError: 'Expr' object has no attribute 'end_lineno'
	```
 
- Finally, is there a way to simplify my code? I don't know all the functions that exist, nor do I know all the `_Py*` functions so I'd be happy if you could tell me if there's a shortcut function for `', '.join(map(repr, sorted(x)))` ....

File: Parser/asdl_c.py [hunk 12]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+            Py_hash_t hash;
+            while (_PySet_NextEntry(optionals, &pos, &item, &hash)) {
+                if(PySet_Discard(expecting, item) < 0) {
+                    Py_DECREF(expecting);
+                    Py_DECREF(optionals);
+                    return -1;
+                }
+            }
+            Py_DECREF(optionals);
+        }
```
Comment (picnixz) [inline Parser/asdl_c.py]: Is there a way for me to simplify this one? Is it fine for me to call a set method directly (like, `.difference`?)

File: Parser/asdl_c.py [hunk 13]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+ * exception and returns -1 (no exception is set if some
+ * other internal errors occur).
+ *
+ * @param self          The AST node instance.
+ * @param dict          The AST node instance dictionary (self.__dict__).
+ * @param fields        The list of fields (self._fields).
+ * @param attributes    The list of attributes (self._attributes).
+ * @param kwargs        Keyword arguments passed to ast_type_replace().
+ *
+ * The @e dict, @e fields, @e attributes and @e kwargs arguments can be NULL.
```
Comment (picnixz) [inline Parser/asdl_c.py]: I tend to comment everything, especially for the future me (or future someone else) but I'm not sure if you want to use doxygen-like comments (note that I did not use `/**` but only `/*` since this may render differently in IDEs and `/**` is almost never used).

If you want me to remove the doxygen directives, I can do it as well (though I think it's nice to have them).

File: Parser/asdl_c.py [hunk 14]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+        if (sep == NULL) {
+            Py_DECREF(names);
+            return -1;
+        }
+        PyObject *str_names = PyUnicode_Join(sep, names);
+        Py_DECREF(sep);
+        Py_DECREF(names);
+        if (str_names == NULL) {
+            return -1;
+        }
```
Comment (picnixz) [inline Parser/asdl_c.py:1272]: Same question here, but do we have some utility function for the Python equivalent:

```py
names = ', '.join(sorted(map(repr, iterable)))
```

If not, would it make sense to provide such functions (only to be used inside CPython)?

**Review by picnixz — COMMENTED** [review]
I think with this implementation we raise exceptions whenever it is needed (except at one point where a warning is emitted, but I think it's a separate issue; see my previous comment).

Anyway I hope that I have made the requested changes; please review again.

**Comment by picnixz** [comment]
Ah the bot did not catch the fact that I said that I was done with the changes.

I have made the requested changes; please review again.

File: Lib/test/test_ast.py [hunk 15]
```diff
@@ -1218,6 +1237,259 @@ def test_copy_with_parents(self):
...
+            y = 1
+            _attributes = ('x', 'y')
+
+        node = MyNode()
+        self.assertEqual(node.x, 0)
+        self.assertEqual(node.y, 1)
+
+        y = object()
+        # custom attributes are currently not supported and raise a warning
+        # because the allowed attributes are hard-coded !
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py:1390]: This seems like we're testing for the presence of a bug. But this is a good catch and I think we need to fix it for 3.13.

I'll open an issue about this since it should definitely be fixed for 3.13.

Comment (picnixz) [reply Lib/test/test_ast.py:1390]: I wasn't really sure whether this was intended or not, and I forgot to ask you that! Since I didn't want to touch the constructor in this PR, I waited for your feedback

File: Parser/asdl_c.py [hunk 16]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
     return result;
 }
 
+/*
+ * Perform the following validations:
+ *
+ *   - All keyword arguments are known @e fields or @e attributes.
```
Comment (JelleZijlstra) [inline Parser/asdl_c.py]: What is "``@e`"? If this is a doxygen-specific convention, I think it's better avoided in this codebase since there's no other C code using it.

File: Parser/asdl_c.py [hunk 17]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+    if (expecting == NULL) {
+        return -1;
+    }
+    if (attributes) {
+        if (_PySet_Update(expecting, attributes) < 0) {
+            Py_DECREF(expecting);
+            return -1;
+        }
+    }
+    // Any keyword argument that is neither a field or attribute is rejected.
```
Comment (JelleZijlstra) [inline Parser/asdl_c.py]:
```suggestion
    // Any keyword argument that is neither a field nor attribute is rejected.
```

File: Parser/asdl_c.py (hunk 12 above)
Comment (JelleZijlstra) [reply Parser/asdl_c.py]: Yes, you can use https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethod or a similar function.

File: Parser/asdl_c.py (hunk 14 above)
Comment (JelleZijlstra) [reply Parser/asdl_c.py:1272]: I think what you have is fine. An alternative approach could be to use PyUnicodeWriter but I don't know if you'll end up with less code.

File: Parser/asdl_c.py [hunk 18]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+            goto cleanup;
+        }
+    }
+    if (kwargs && PyDict_Update(payload, kwargs) < 0) {
+        goto cleanup;
+    }
+    result = PyObject_Call(type, empty_tuple, payload);
+    if (result == NULL) {
+        goto cleanup;
+    }
```
Comment (JelleZijlstra) [inline Parser/asdl_c.py]:
```suggestion
```
You'll go there anyway.

File: Parser/asdl_c.py [hunk 19]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+ * @param self          The AST node instance.
+ * @param dict          The AST node instance dictionary (self.__dict__).
+ * @param fields        The list of fields (self._fields).
+ * @param attributes    The list of attributes (self._attributes).
+ * @param kwargs        Keyword arguments passed to ast_type_replace().
+ *
+ * The @e dict, @e fields, @e attributes and @e kwargs arguments can be NULL.
+ */
+static inline int
+ast_type_replace_check(PyObject *self,
```
Comment (JelleZijlstra) [inline Parser/asdl_c.py:1159]: Probably worth noting that this can be removed in 3.15, when the deprecations trigger and we can just rely on the constructor to do these checks.

File: Parser/asdl_c.py [hunk 20]
```diff
@@ -1132,13 +1132,291 @@ def visitModule(self, mod):
...
+            PyObject *optionals = PySet_New(attributes);
+            if (optionals == NULL) {
+                Py_DECREF(expecting);
+                return -1;
+            }
+            Py_ssize_t pos = 0;
+            PyObject *item;
+            Py_hash_t hash;
+            while (_PySet_NextEntry(optionals, &pos, &item, &hash)) {
+                if(PySet_Discard(expecting, item) < 0) {
```
Comment (JelleZijlstra) [inline Parser/asdl_c.py]:
```suggestion
                if (PySet_Discard(expecting, item) < 0) {
```

File: Parser/asdl_c.py (hunk 16 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Ah, it's the general way to format parameters in doxygen: https://www.doxygen.nl/manual/commands.html#cmde

That's what I wanted to ask but do you want me to use non-doxygen comments in general?

Comment (picnixz) [reply Parser/asdl_c.py]: Actually, I remember that I need `/**`. But that would be how it is rendered:

image

But I think I'll remove those if it's maybe too much?

Comment (JelleZijlstra) [reply Parser/asdl_c.py]: Yes, I think these doxygen-specific conventions are too much since the codebase doesn't generally use them.

File: Parser/asdl_c.py (hunk 14 above)
Comment (picnixz) [reply Parser/asdl_c.py:1272]: I'll keep this one then. I still need to construct that sorted list (maybe you have a shortcut for `sorted(map(repr, iterable))`? if not, then I'll use the current code.

File: Parser/asdl_c.py (hunk 12 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Maybe you can explain it to me but why:

```C
PyObject *unused = PyObject_CallMethod(expecting,
                                       "difference_update",
									   "O",
                                       attributes);
```

does nothing but

```C
PyObject *unused = PyObject_CallMethodOneArg(expecting,
                                             &_Py_ID(difference_update),
                                             attributes);
```

does what I expected? (I'm pretty I'm incorrectly using `PyObject_CallMethod` but the call seems to succeed (it doesn't crash)).

File: Parser/asdl_c.py (hunk 14 above)
Comment (JelleZijlstra) [reply Parser/asdl_c.py:1272]: I'd probably do PySequence_List, then PyList_Sort, then iterate over the list and add to the writer.

Comment (JelleZijlstra) [reply Parser/asdl_c.py:1272]: (But again, not convinced that's going to be simpler than what you have, and it does add an extra intermediate list.)

**Review by JelleZijlstra — APPROVED** [review]

File: Parser/asdl_c.py (hunk 12 above)
Comment (JelleZijlstra) [reply Parser/asdl_c.py]: I think you need `"(O)"`, not `"O"`.

As for why it silently does the wrong thing, well, it's C :)

File: Parser/asdl_c.py (hunk 14 above)
Comment (picnixz) [reply Parser/asdl_c.py:1272]: Yes. Maybe I'll suggest some helper function but I'll stick with what I did here. I prefer not to allocate too many temporary objects...

File: Parser/asdl_c.py (hunk 12 above)
Comment (picnixz) [reply Parser/asdl_c.py]: Oh thank you. So I needed another format. I expected there would be some safeguards (namely, the result would be NULL if I call it incorrectly) but maybe I wished for too much :)

**Comment by picnixz** [comment]
(I hope I didn't leak things. The previous implementation didn't but maybe this one does). I'll be leaving for today, my defense is tomorrow so I'll patch if afterwards if needed.

**Comment by JelleZijlstra** [comment]
Just running the refleak bots because it's good practice to do that when there's a big pile of new C code.

Good luck with your defense!
